File: wiii-desktop/src/lib/pointy-fast-path.ts
import type { HostContext } from "@/stores/host-context-store";

export const POINTY_FAST_PATH_SOURCE = "pointy_fast_path";

const LOCATE_TERMS = [
  "o dau",
  "where",
  "chi",
  "chi cho",
  "tim",
  "tro",
  "highlight",
  "nut",
  "button",
];

const CLICK_TERMS = [
  "mo",
  "bam",
  "click",
  "nhan vao",
  "vao",
  "di toi",
  "chuyen toi",
  "open",
];

const UNSAFE_CLICK_TERMS = [
  "submit",
  "nop bai",
  "quiz",
  "checkout",
  "payment",
  "thanh toan",
  "enroll",
  "dang ky",
  "logout",
  "dang xuat",
  "delete",
  "xoa",
  "mark complete",
  "hoan thanh",
];

const TARGET_ALIASES: Array<{ ids: string[]; terms: string[] }> = [
  {
    ids: ["browse-courses", "browse-courses-link", "browse-courses-button"],
    terms: ["kham pha khoa hoc", "kham pha", "browse courses", "browse"],
  },
  {
    ids: ["continue-learn", "continue-lesson", "continue-course"],
    terms: ["tiep tuc hoc", "tiep tuc", "continue learning", "continue"],
  },
  {
    ids: ["my-courses", "my-courses-link"],
    terms: ["khoa hoc cua toi", "khoa hoc dang hoc", "my courses"],
  },
  {
    ids: ["profile-link", "profile-card"],
    terms: ["ho so", "profile"],
  },
];

export interface PointyFastPathTarget {
  id: string;
  selector: string;
  label?: string;
  click_safe?: boolean;
  click_kind?: string;
}

export interface PointyFastPathAction {
  action: "ui.highlight" | "ui.click";
  requestId: string;
  params: {
    selector: string;
    message?: string;
    duration_ms?: number;
    source: typeof POINTY_FAST_PATH_SOURCE;
  };
  target: PointyFastPathTarget;
  reason: "locate" | "click" | "unsafe_click_demoted";
}

export function normalizePointyText(value: string): string {
  return value
    .replace(/đ/g, "d")
    .replace(/Đ/g, "d")
    .normalize("NFD")
    .replace(/\p{Diacritic}/gu, "")
    .replace(/đ/g, "d")
    .replace(/Đ/g, "d")
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, " ")
    .trim()
    .replace(/\s+/g, " ");
}

function hasAnyTerm(prompt: string, terms: string[]): boolean {
  return terms.some((term) => prompt.includes(term));
}

function coerceTarget(value: unknown): PointyFastPathTarget | null {
  if (!value || typeof value !== "object") return null;
  const raw = value as Record<string, unknown>;
  const id = typeof raw.id === "string" ? raw.id.trim() : "";
  const selector = typeof raw.selector === "string" ? raw.selector.trim() : "";
  if (!id || !selector) return null;
  return {
    id,
    selector,
    label: typeof raw.label === "string" ? raw.label.trim() : undefined,
    click_safe: raw.click_safe === true,
    click_kind: typeof raw.click_kind === "string" ? raw.click_kind.trim() : undefined,
  };
}

export function getPointyTargetsFromContext(ctx: HostContext | null): PointyFastPathTarget[] {
  const targets = ctx?.page?.metadata?.available_targets;
  if (!Array.isArray(targets)) return [];
  return targets.map(coerceTarget).filter((target): target is PointyFastPathTarget => Boolean(target));
}

function aliasScore(prompt: string, target: PointyFastPathTarget): number {
  const idText = normalizePointyText(target.id);
  for (const alias of TARGET_ALIASES) {
    const aliasMatchesPrompt = alias.terms.some((term) => prompt.includes(term));
    if (!aliasMatchesPrompt) continue;
    if (alias.ids.some((id) => idText.includes(normalizePointyText(id)))) {
      return 40;
    }
  }
  return 0;
}

function targetScore(prompt: string, target: PointyFastPathTarget): number {
  const idText = normalizePointyText(target.id);
  const selectorText = normalizePointyText(target.selector);
  const labelText = normalizePointyText(target.label || "");
  const labelWords = labelText.split(" ").filter((word) => word.length >= 3);
  let score = aliasScore(prompt, target);

  if (labelText && prompt.includes(labelText)) score += 35;
  if (idText && prompt.includes(idText)) score += 24;
  if (selectorText && prompt.includes(selectorText)) score += 12;
  if (labelWords.length > 0) {
    const matchedWords = labelWords.filter((word) => prompt.includes(word)).length;
    score += matchedWords * 6;
    if (matchedWords === labelWords.length) score += 12;
  }

  if (target.click_safe) score += 2;
  return score;
}

function selectTarget(prompt: string, targets: PointyFastPathTarget[]): PointyFastPathTarget | null {
  let best: { target: PointyFastPathTarget; score: number } | null = null;
  for (const target of targets) {
    const score = targetScore(prompt, target);
    if (score <= 0) continue;
    if (!best || score > best.score) {
      best = { target, score };
    }
  }
  return best && best.score >= 12 ? best.target : null;
}

function isUnsafeClickTarget(prompt: string, target: PointyFastPathTarget): boolean {
  const combined = normalizePointyText(
    `${prompt} ${target.id} ${target.label || ""} ${target.click_kind || ""}`,
  );
  return hasAnyTerm(combined, UNSAFE_CLICK_TERMS);
}

function makeRequestId(): string {
  const random = globalThis.crypto?.randomUUID?.() || Math.random().toString(36).slice(2, 12);
  return `pointy-fast-${random}`;
}

export function buildPointyFastPathAction(
  prompt: string,
  ctx: HostContext | null,
): PointyFastPathAction | null {
  const normalizedPrompt = normalizePointyText(prompt);
  if (!normalizedPrompt) return null;
  const wantsLocate = hasAnyTerm(normalizedPrompt, LOCATE_TERMS);
  const wantsClick = hasAnyTerm(normalizedPrompt, CLICK_TERMS);
  if (!wantsLocate && !wantsClick) return null;

  const target = selectTarget(normalizedPrompt, getPointyTargetsFromContext(ctx));
  if (!target) return null;

  const label = target.label || target.id;
  if (wantsClick && target.click_safe && !isUnsafeClickTarget(normalizedPrompt, target)) {
    return {
      action: "ui.click",
      requestId: makeRequestId(),
      params: {
        selector: target.id,
        message: `Wiii dang mo ${label} cho ban.`,
        source: POINTY_FAST_PATH_SOURCE,
      },
      target,
      reason: "click",
    };
  }

  return {
    action: "ui.highlight",
    requestId: makeRequestId(),
    params: {
      selector: target.id,
      message: `Day la ${label}. Wiii tro vao de ban thay ngay.`,
      duration_ms: wantsClick ? 5600 : 5200,
      source: POINTY_FAST_PATH_SOURCE,
    },
    target,
    reason: wantsClick ? "unsafe_click_demoted" : "locate",
  };
}

File: maritime-ai-service/app/engine/context/pointy_fast_path.py
"""Fast local Pointy intent matching for host UI guidance.

This is intentionally conservative. It exists to emit safe UI guidance before
the slower model/router path decides whether to call a host action tool.
"""
from __future__ import annotations

import re
import unicodedata
import uuid
from typing import Any

from app.engine.context.pointy_actions import POINTY_ACTION_CLICK, POINTY_ACTION_HIGHLIGHT

POINTY_FAST_PATH_SOURCE = "pointy_fast_path"

_LOCATE_TERMS = (
    "o dau",
    "where",
    "chi",
    "chi cho",
    "tim",
    "tro",
    "highlight",
    "nut",
    "button",
)
_CLICK_TERMS = (
    "mo",
    "bam",
    "click",
    "nhan vao",
    "vao",
    "di toi",
    "chuyen toi",
    "open",
)
_UNSAFE_CLICK_TERMS = (
    "submit",
    "nop bai",
    "quiz",
    "checkout",
    "payment",
    "thanh toan",
    "enroll",
    "dang ky",
    "logout",
    "dang xuat",
    "delete",
    "xoa",
    "mark complete",
    "hoan thanh",
)

_TARGET_ALIASES: tuple[tuple[tuple[str, ...], tuple[str, ...]], ...] = (
    (
        ("browse-courses", "browse-courses-link", "browse-courses-button"),
        ("kham pha khoa hoc", "kham pha", "browse courses", "browse"),
    ),
    (
        ("continue-learn", "continue-lesson", "continue-course"),
        ("tiep tuc hoc", "tiep tuc", "continue learning", "continue"),
    ),
    (
        ("my-courses", "my-courses-link"),
        ("khoa hoc cua toi", "khoa hoc dang hoc", "my courses"),
    ),
    (
        ("profile-link", "profile-card"),
        ("ho so", "profile"),
    ),
)


def normalize_pointy_text(value: str) -> str:
    normalized = (value or "").replace("đ", "d").replace("Đ", "d")
    normalized = unicodedata.normalize("NFD", normalized)
    normalized = "".join(ch for ch in normalized if unicodedata.category(ch) != "Mn")
    normalized = normalized.replace("đ", "d").replace("Đ", "d").lower()
    normalized = re.sub(r"[^a-z0-9]+", " ", normalized)
    return re.sub(r"\s+", " ", normalized).strip()


def _has_any_term(prompt: str, terms: tuple[str, ...]) -> bool:
    return any(term in prompt for term in terms)


def _pointy_fast_path_already_ran(context: dict[str, Any] | None) -> bool:
    feedback = (context or {}).get("host_action_feedback")
    if not isinstance(feedback, dict):
        return False
    last = feedback.get("last_action_result")
    if not isinstance(last, dict):
        return False
    params = last.get("params")
    return isinstance(params, dict) and params.get("source") == POINTY_FAST_PATH_SOURCE


def _coerce_target(value: Any) -> dict[str, Any] | None:
    if not isinstance(value, dict):
        return None
    target_id = str(value.get("id") or "").strip()
    selector = str(value.get("selector") or "").strip()
    if not target_id or not selector:
        return None
    return {
        "id": target_id,
        "selector": selector,
        "label": str(value.get("label") or "").strip() or None,
        "click_safe": value.get("click_safe") is True,
        "click_kind": str(value.get("click_kind") or "").strip() or None,
    }


def get_pointy_targets_from_context(context: dict[str, Any] | None) -> list[dict[str, Any]]:
    ctx = context or {}
    host_context = ctx.get("host_context") if isinstance(ctx.get("host_context"), dict) else {}
    host_page = host_context.get("page") if isinstance(host_context.get("page"), dict) else {}
    host_metadata = host_page.get("metadata") if isinstance(host_page.get("metadata"), dict) else {}
    page_context = ctx.get("page_context") if isinstance(ctx.get("page_context"), dict) else {}
    raw_targets = host_metadata.get("available_targets")
    if raw_targets is None:
        raw_targets = page_context.get("available_targets")
    if not isinstance(raw_targets, list):
        return []
    return [target for item in raw_targets if (target := _coerce_target(item))]


def _alias_score(prompt: str, target: dict[str, Any]) -> int:
    target_id = normalize_pointy_text(str(target.get("id") or ""))
    for alias_ids, alias_terms in _TARGET_ALIASES:
        if not any(term in prompt for term in alias_terms):
            continue
        if any(normalize_pointy_text(alias_id) in target_id for alias_id in alias_ids):
            return 40
    return 0


def _target_score(prompt: str, target: dict[str, Any]) -> int:
    target_id = normalize_pointy_text(str(target.get("id") or ""))
    selector = normalize_pointy_text(str(target.get("selector") or ""))
    label = normalize_pointy_text(str(target.get("label") or ""))
    label_words = [word for word in label.split(" ") if len(word) >= 3]

    score = _alias_score(prompt, target)
    if label and label in prompt:
        score += 35
    if target_id and target_id in prompt:
        score += 24
    if selector and selector in prompt:
        score += 12
    if label_words:
        matched = sum(1 for word in label_words if word in prompt)
        score += matched * 6
        if matched == len(label_words):
            score += 12
    if target.get("click_safe"):
        score += 2
    return score


def _select_target(prompt: str, targets: list[dict[str, Any]]) -> dict[str, Any] | None:
    best_target: dict[str, Any] | None = None
    best_score = 0
    for target in targets:
        score = _target_score(prompt, target)
        if score > best_score:
            best_target = target
            best_score = score
    return best_target if best_score >= 12 else None


def _is_unsafe_click_target(prompt: str, target: dict[str, Any]) -> bool:
    combined = normalize_pointy_text(
        f"{prompt} {target.get('id') or ''} {target.get('label') or ''} {target.get('click_kind') or ''}"
    )
    return _has_any_term(combined, _UNSAFE_CLICK_TERMS)


def build_pointy_fast_path_action(
    prompt: str,
    context: dict[str, Any] | None,
) -> dict[str, Any] | None:
    if _pointy_fast_path_already_ran(context):
        return None

    normalized_prompt = normalize_pointy_text(prompt)
    wants_locate = _has_any_term(normalized_prompt, _LOCATE_TERMS)
    wants_click = _has_any_term(normalized_prompt, _CLICK_TERMS)
    if not normalized_prompt or (not wants_locate and not wants_click):
        return None

    target = _select_target(normalized_prompt, get_pointy_targets_from_context(context))
    if not target:
        return None

    label = str(target.get("label") or target["id"])
    if wants_click and target.get("click_safe") and not _is_unsafe_click_target(normalized_prompt, target):
        return {
            "request_id": f"pointy-fast-{uuid.uuid4()}",
            "action": POINTY_ACTION_CLICK,
            "params": {
                "selector": target["id"],
                "message": f"Wiii đang mở {label} cho bạn.",
                "source": POINTY_FAST_PATH_SOURCE,
            },
            "target": target,
            "reason": "click",
        }

    return {
        "request_id": f"pointy-fast-{uuid.uuid4()}",
        "action": POINTY_ACTION_HIGHLIGHT,
        "params": {
            "selector": target["id"],
            "message": f"Đây là {label}. Wiii trỏ vào để bạn thấy ngay.",
            "duration_ms": 5600 if wants_click else 5200,
            "source": POINTY_FAST_PATH_SOURCE,
        },
        "target": target,
        "reason": "unsafe_click_demoted" if wants_click else "locate",
    }


__all__ = [
    "POINTY_FAST_PATH_SOURCE",
    "build_pointy_fast_path_action",
    "get_pointy_targets_from_context",
    "normalize_pointy_text",
]
